worker/multiwatcher/watcher.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package multiwatcher
+
+import (
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/multiwatcher"
+)
+
+type control interface {
+	Dying() <-chan struct{}
+	Err() error
+}
+
+// Watcher watches any changes to the state.
+type Watcher struct {
+	request chan *request
+	control control
+	logger  Logger
+	err     chan error
+
+	filter func([]multiwatcher.Delta) []multiwatcher.Delta
+
+	// The following fields are maintained by the Worker goroutine.
+	revno   int64
+	stopped bool
+
+	// used indicates that the watcher was used (i.e. Next() called).
+	used bool
+}
+
+// Stop stops the watcher.
+func (w *Watcher) Stop() error {
+	select {
+	case w.request <- &request{watcher: w}:
+		// We asked to be stopped, and this is processed.
+	case _, stillOpen := <-w.err:
+		// We have been stopped already.
+		if stillOpen {
+			close(w.err)
+		}
+		// Since we are asking to stop, we don't care about the underlying
+		// error from the read.
+	case <-w.control.Dying():
+		// The worker is dying and we will be cleaned up.
+	}
+	// In all of the above cases, we're fine, and don't need to return an
+	// error that would be logged.
+	return nil
+}
+
+// Next retrieves all changes that have happened since the last
+// time it was called, blocking until there are some changes available.
+//
+// The result from the initial call to Next() is different from
+// subsequent calls. The latter will reflect changes that have happened
+// since the last Next() call. In contrast, the initial Next() call will
+// return the deltas that represent the model's complete state at that
+// moment, even when the model is empty. In that empty model case an
+// empty set of deltas is returned.
+func (w *Watcher) Next() ([]multiwatcher.Delta, error) {
+	// In order to be able to apply the filter, yet not signal the caller when
+	// all deltas were filtered out, we need an outer loop.
+	var changes []multiwatcher.Delta
+	for len(changes) == 0 {
+		req := &request{
+			watcher: w,
+			reply:   make(chan bool),
+		}
+		if !w.used {
```

Comment:
Probably a silly question, but does Next() get called from multiple goroutines?

Reply:
The Next() calls come from the apiserver generated goroutines. So very independent from the goroutines of the worker responsible for the underlying changes.